src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -13,730 +13,874 @@ use rustc_mir::monomorphize;
+        // Create the callee. This is a fn ptr or zero-sized and hence a kind of scalar.
+        let callee = self.codegen_operand(&mut bx, func);
+
+        let (instance, mut llfn) = match callee.layout.ty.sty {
+            ty::FnDef(def_id, substs) => {
+                (Some(ty::Instance::resolve(bx.tcx(),
+                                            ty::ParamEnv::reveal_all(),
+                                            def_id,
+                                            substs).unwrap()),
+                 None)
+            }
+            ty::FnPtr(_) => {
+                (None, Some(callee.immediate()))
+            }
+            _ => bug!("{} is not callable", callee.layout.ty),
+        };
+        let def = instance.map(|i| i.def);
+        let sig = callee.layout.ty.fn_sig(bx.tcx());
+        let sig = bx.tcx().normalize_erasing_late_bound_regions(
+            ty::ParamEnv::reveal_all(),
+            &sig,
+        );
+        let abi = sig.abi;
+
+        // Handle intrinsics old codegen wants Expr's for, ourselves.
+        let intrinsic = match def {
+            Some(ty::InstanceDef::Intrinsic(def_id)) =>
+                Some(bx.tcx().item_name(def_id).as_str()),
+            _ => None
+        };
+        let intrinsic = intrinsic.as_ref().map(|s| &s[..]);
 
-                // Create the failure block and the conditional branch to it.
-                let lltarget = llblock(self, target);
-                let panic_block = self.new_block("panic");
-                if expected {
-                    bx.cond_br(cond, lltarget, panic_block.llbb());
-                } else {
-                    bx.cond_br(cond, panic_block.llbb(), lltarget);
-                }
+        if intrinsic == Some("transmute") {
+            if let Some(destination_ref) = destination.as_ref() {
+                let &(ref dest, target) = destination_ref;
+                self.codegen_transmute(&mut bx, &args[0], dest);
+                helper.funclet_br(self, &mut bx, target);
+            } else {
+                // If we are trying to transmute to an uninhabited type,
+                // it is likely there is no allotted destination. In fact,
+                // transmuting to an uninhabited type is UB, which means
+                // we can do what we like. Here, we declare that transmuting
+                // into an uninhabited type is impossible, so anything following
+                // it must be unreachable.
+                assert_eq!(bx.layout_of(sig.output()).abi, layout::Abi::Uninhabited);
+                bx.unreachable();
+            }
+            return;
+        }
 
-                // After this point, bx is the block for the call to panic.
-                bx = panic_block;
-                self.set_debug_loc(&mut bx, terminator.source_info);
+        // The "spoofed" VaList added to a C variadic functions signature
+        // should not be included in the extra_args calculation.
+        let extra_args = if sig.variadic {
```

Comment:
I think this would be better...
```rust
let extra_args_start_idx = sig.inputs().len() - if sig.variadic { 1 } else { 0 };
let extra_args = &args[extra_args_start_idx..];
```